### PR TITLE
Fitness project initialization

### PR DESCRIPTION
**Pull request by olgamazyuk**
Fitness project initialization

### CODE REVIEW

File: client/src/pages/AuthPage.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    password: "",
+  });
+
+  useEffect(() => {
+    message(error);
+    clearError();
+  }, [error, message, clearError]);
+
+  useEffect(() => {
+    window.M.updateTextFields();
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js:21]: it is a function from Materialize which is used when labels overlap input field. After using this function label is before input but not inside it overlapping placeholder. But I decided not to use labels here, so I don't need this function.

File: client/src/pages/ProfilePage.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+export const ProfilePage = () => {
+  return (
+    <div className="grey lighten-4 profile-container">
+      <h3 className="text-center">Мой кабинет</h3>
+      <form action="#" className="mb-40">
+        <h5>Ваш пол:</h5>
+        <label className="mr-10">
+          <input className="with-gap" type="radio" />
+          <span>Мужской</span>
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/ProfilePage.js]: fixed

File: client/src/pages/HistoryPage.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import { useHttp } from "../hooks/http.hook";
+import { Loader } from "../components/Loader";
+
+export const HistoryPage = () => {
+  const { loading } = useHttp();
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/HistoryPage.js:6]: yes

File: client/src/pages/HistoryPage.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+  if (loading) {
+    return <Loader />;
+  }
+
+  return <>{!loading && <p className="center">Информации пока что нет</p>}</>;
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/HistoryPage.js:12]: fixed

File: client/src/pages/AuthPage.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }, []);
+
+  const changeHandler = (event) => {
+    setForm({ ...form, [event.target.name]: event.target.value });
+  };
+
+  const registerHandler = async () => {
+    try {
+      const data = await request("/api/auth/register", "POST", { ...form });
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js]: I thought that I deal with all http requests in client/src/hooks/http.hook.js
fixed api routes

File: client/src/pages/AddCaloriesPage.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+          </select>
+        </div>
+        <div className="input-field row s6">
+          <input id="quantity" type="text" className="validate" />
+          <label htmlFor="quantity">Количество в граммах</label>
+        </div>
+        <button className="btn">Добавить</button>
+      </div>
+      <div className="col s6 p-100">
+        <h3>Тренировки</h3>
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/AddCaloriesPage.js]: fixed

File: client/src/hooks/auth.hook.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    localStorage.removeItem(storageName);
+  }, []);
+
+  useEffect(() => {
+    const data = JSON.parse(localStorage.getItem(storageName));
+
+    if (data && data.token) {
+      login(data.token, data.userId);
+    }
+    setReady(true);
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/hooks/auth.hook.js:35]: fixed

File: client/src/context/AuthContext.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { createContext } from "react";
+
+function noop() {}
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/context/AuthContext.js]: fixed

File: client/src/components/Navbar.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    history.push("/");
+  };
+
+  return (
+    <nav>
+      <div className="nav-wrapper orange lighten-1 navbar-container">
+        <span className="brand-logo">Подсчет калорий</span>
+        <ul id="nav-mobile" className="right hide-on-med-and-down">
+          <li>
+            <NavLink to="/profile">Кабинет пользователя</NavLink>
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/components/Navbar.js:21]: fixed

File: client/src/hooks/auth.hook.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import { useState, useCallback, useEffect } from "react";
+
+const storageName = "userData";
+
+export const useAuth = () => {
+  const [token, setToken] = useState(null);
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/hooks/auth.hook.js:6]: fixed

File: client/src/hooks/auth.hook.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { useState, useCallback, useEffect } from "react";
+
+const storageName = "userData";
```
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/hooks/auth.hook.js:3]: fixed

File: client/src/components/CategoriesList.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import React from "react";
+
+export const CategoriesList = ({ categories }) => {
+  return (
+    <form className="row">
+      <div className="col s6 p-100">
+        <h3>Еда и напитки</h3>
```
Comment (nikolyachikm) [inline client/src/components/CategoriesList.js]: can we please separate such things? 
Food and trainings are different things by logic, so better to have separated components for them

File: client/src/components/CategoriesList.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        <div className="row green-text">
+          <label>Выберите продукт</label>
+          <select className="browser-default">
+            {categories
+              .filter((el) => el.type === 1)
```
Comment (nikolyachikm) [inline client/src/components/CategoriesList.js]: we should just render values in views (except maybe some edge cases), so move please all needed filtration to the service that provides list of categories. 
1 here is magic number - use something like enums please for fields like `type`

File: client/src/pages/AddCaloriesPage.js [hunk 14]
```diff
@@ -1,50 +1,35 @@
...
-        </div>
-        <button className="btn">Добавить</button>
-      </div>
-    </form>
-  );
+  const [categories, setCategories] = useState([]);
+  const { loading, request } = useHttp();
+  const { token } = useContext(AuthContext);
+
+  const getCategories = useCallback(async () => {
```
Comment (nikolyachikm) [inline client/src/pages/AddCaloriesPage.js:14]: create categoriesService please and here we should just call `categoriesService.getCategories()` or maybe 
const foodCategories = categoriesService.getFoodCategories(); 
const trainingCategories = categoriesService.getTrainingCategories();

File: client/src/pages/AddCaloriesPage.js [hunk 15]
```diff
@@ -1,50 +1,35 @@
...
+          Authorization: `Bearer ${token}`,
+        }
+      );
+      setCategories(fetched);
+    } catch (e) {}
+  }, [token, request]);
+
+  useEffect(() => {
+    getCategories();
+  }, [getCategories]);
```
Comment (nikolyachikm) [inline client/src/pages/AddCaloriesPage.js]: i think here we should have just []

File: models/Data.js [hunk 16]
```diff
@@ -0,0 +1,10 @@
+const { Schema, model, Types } = require("mongoose");
+
+const schema = new Schema({
+  gender: { type: String, required: true },
+  activity: { type: String, required: true },
+  age: { type: String, required: true },
+  owner: { type: Types.ObjectId, ref: "User" },
```
Comment (nikolyachikm) [inline models/Data.js]: do we really need it? i think we can just use User model to access Data

File: client/src/pages/ProfilePage.js [hunk 17]
```diff
@@ -1,67 +1,173 @@
...
+    age: "",
+  });
+
+  const changeHandler = (event) => {
+    setUserData({ ...userData, [event.target.name]: event.target.value });
+  };
+
+  const submitHandler = async () => {
+    try {
+      const data = await request(
```
Comment (nikolyachikm) [inline client/src/pages/ProfilePage.js:23]: move to profileService or similar

main idea that we should not mix some api actions with view. making requests, preparing data, filtering response,etc - that's service's responsibility.
view should just render data. call redux actions/ or services. if we mix all things, code will be not readable in real project

File: client/src/pages/AuthPage.js [hunk 18]
```diff
@@ -27,14 +27,22 @@ export const AuthPage = () => {
...
+        { ...form }
+      );
       message(data.message);
     } catch (e) {}
   };
 
   const loginHandler = async () => {
     try {
-      const data = await request("/api/auth/login", "POST", { ...form });
+      const data = await request(
```
Comment (nikolyachikm) [inline client/src/pages/AuthPage.js]: the same as below - move to service please

File: client/src/pages/AuthPage.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikolyachikm) [reply client/src/pages/AuthPage.js]: in project we have many api calls usually. often we need to amend data before sending request or filter data in response for example. so we can't move such code into single file, or keep in views. we need to create separate services (categories - one file, user profile - another file, etc.). this is keeping code clear, maintainable, testable, and so on))

[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js]: fixed

File: client/src/components/CategoriesList.js (hunk 12 above)
Comment (olgamazyuk) [reply client/src/components/CategoriesList.js]: fixed

File: client/src/components/CategoriesList.js (hunk 13 above)
Comment (olgamazyuk) [reply client/src/components/CategoriesList.js]: fixed

File: client/src/pages/AddCaloriesPage.js (hunk 14 above)
Comment (olgamazyuk) [reply client/src/pages/AddCaloriesPage.js:14]: fixed

File: client/src/pages/AddCaloriesPage.js (hunk 15 above)
Comment (olgamazyuk) [reply client/src/pages/AddCaloriesPage.js]: fixed

File: client/src/pages/AuthPage.js (hunk 18 above)
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js]: fixed

File: client/src/pages/AuthPage.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js]: fixed

File: models/Data.js (hunk 16 above)
Comment (olgamazyuk) [reply models/Data.js]: agree, fixed

File: client/src/components/TrainingList.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { trainingData } from '../enum';
+
+export const TrainingList = ({ training }) => {
```
Comment (nikolyachikm) [inline client/src/components/TrainingList.js:4]: thids is duplication of product list - can we make 1 common view and just pass different entities in parent view ?

File: client/src/hooks/auth.hook.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+  useEffect(() => {
+    const data = JSON.parse(localStorage.getItem(STORAGE_NAME));
+
+    if (data && data.token) {
+      login(data.token, data.userId);
+    }
+
+    setReady(true);
+  }, [ login ]);
```
Comment (nikolyachikm) [inline client/src/hooks/auth.hook.js:37]: maybe [] is enough here?

File: client/src/hooks/http.hook.js [hunk 21]
```diff
@@ -0,0 +1,38 @@
+import { useState, useCallback } from 'react';
+
+export const useHttp = () => {
+  const [ loading, setLoading ] = useState(false);
+  const [ error, setError ] = useState(null);
+
+  const request = useCallback(
+    async (url, method = 'GET', body = null, headers = {}) => {
+      setLoading(true);
+      try {
```
Comment (nikolyachikm) [inline client/src/hooks/http.hook.js:10]: new lines between code blocks please

also you can just use axios and it'll make code simplier

File: client/src/http/AuthService.js [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    }
+  );
+};
+
+export const loginHandle = async (form) => {
+  return await axios.post(
+    paths.Login,
+    { ...form },
+    {
+      headers: {
```
Comment (nikolyachikm) [inline client/src/http/AuthService.js:21]: get rid of code duplication please

File: client/src/http/CategoriesService.js [hunk 23]
```diff
@@ -0,0 +1,10 @@
+import axios from 'axios';
+import { paths } from '../enum';
+
+export const getItems = async (token) => {
+  return await axios.get(paths.Categories, {
+    headers: {
```
Comment (nikolyachikm) [inline client/src/http/CategoriesService.js:6]: the same

File: client/src/pages/AddCaloriesPage.js [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+  const [ products, setProducts ] = useState([]);
+  const [ training, setTraining ] = useState([]);
+  const { loading } = useHttp();
+
+  const getCategories = async () => {
+    try {
+      const { data } = await getItems(token);
+      setProducts(data.products);
+      setTraining(data.training);
+    } catch (e) {}
```
Comment (nikolyachikm) [inline client/src/pages/AddCaloriesPage.js:18]: why we need empty catch ? maybe we can just remove it ?

File: client/src/pages/AuthPage.js [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+  const { loading, error, clearError } = useHttp();
+  const [ form, setForm ] = useState({
+    email: '',
+    password: '',
+  });
+
+  useEffect(() => {
+    message(error);
+    clearError();
+  }, [ error, message, clearError ]);
```
Comment (nikolyachikm) [inline client/src/pages/AuthPage.js:20]: just [error]  ?

File: client/src/pages/ProfilePage.js [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+    activity: '',
+    age: '',
+  });
+
+  const changeHandler = (event) => {
+    setUserData({ ...userData, [ event.target.name ]: event.target.value });
+  };
+
+  const submitHandler = async () => {
+    try {
```
Comment (nikolyachikm) [inline client/src/pages/ProfilePage.js:19]: try avoid try catch where it's ot really needed. here we can add interceptor in axios to catch errors and show user messages if needed for example

Comment (nikolyachikm) [reply client/src/pages/ProfilePage.js:19]: and eve forcurrent case - it's weird if user submtted a form, the request failed and no error message will be shown

File: client/src/pages/ProfilePage.js [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+              type="radio"
+              name="gender"
+              value={profileData.GenderMale}
+              onChange={changeHandler}
+            />
+            <span>{profileData.GenderMale}</span>
+          </label>
+          <label>
+            <input
+              className="with-gap"
```
Comment (nikolyachikm) [inline client/src/pages/ProfilePage.js:43]: code duplication :)

File: client/src/pages/ProfilePage.js [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+          </label>
+          <label className="mr-10">
+            <input
+              className="with-gap"
+              type="radio"
+              name="activity"
+              value={profileData.ActivityAverage}
+              onChange={changeHandler}
+            />
+            <span>{profileData.ActivityAverage}</span>
```
Comment (nikolyachikm) [inline client/src/pages/ProfilePage.js:72]: the same

File: client/src/routes.js [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+import { AddCaloriesPage } from './pages/AddCaloriesPage';
+import { ProfilePage } from './pages/ProfilePage';
+import { AuthPage } from './pages/AuthPage';
+import { paths } from './enum';
+import { AuthContext } from './context/AuthContext';
+
+export const useRoutes = (isAuthenticated) => {
+  if (isAuthenticated) {
+    return (
+      <AuthContext.Consumer>
```
Comment (nikolyachikm) [inline client/src/routes.js:13]: move to anothe function please

File: client/src/pages/ProfilePage.js (hunk 27 above)
Comment (nikolyachikm) [reply client/src/pages/ProfilePage.js:43]: ah i see you fixed it in 2nd PR 👍

File: client/src/pages/ProfilePage.js (hunk 28 above)
Comment (nikolyachikm) [reply client/src/pages/ProfilePage.js:72]: fixed it in 2nd PR 👍

File: client/src/components/TrainingList.js (hunk 19 above)
Comment (olgamazyuk) [reply client/src/components/TrainingList.js:4]: fixed

File: client/src/hooks/auth.hook.js (hunk 20 above)
Comment (olgamazyuk) [reply client/src/hooks/auth.hook.js:37]: fixed

File: client/src/hooks/http.hook.js (hunk 21 above)
Comment (olgamazyuk) [reply client/src/hooks/http.hook.js:10]: fixed

File: client/src/http/AuthService.js (hunk 22 above)
Comment (olgamazyuk) [reply client/src/http/AuthService.js:21]: fixed

File: client/src/http/CategoriesService.js (hunk 23 above)
Comment (olgamazyuk) [reply client/src/http/CategoriesService.js:6]: here I have another method (get). Should I fix this anyway?

File: client/src/pages/AddCaloriesPage.js (hunk 24 above)
Comment (olgamazyuk) [reply client/src/pages/AddCaloriesPage.js:18]: fixed

File: client/src/pages/AuthPage.js (hunk 25 above)
Comment (olgamazyuk) [reply client/src/pages/AuthPage.js:20]: fixed

File: client/src/pages/ProfilePage.js (hunk 26 above)
Comment (olgamazyuk) [reply client/src/pages/ProfilePage.js:19]: fixed

File: client/src/routes.js (hunk 29 above)
Comment (olgamazyuk) [reply client/src/routes.js:13]: fixed

**Pull request merged**